### DISCUSSION TITLE
ArrayBuffer tweaks

### DISCUSSION THREAD

**Discussion by japasc** [dotnet/runtime#74107] [General] [answered]
I cant create a PR, sorry.
Two observations

* In method ConfigurableArrayPool.Return
```
if (clearArray)
                {
                    Array.Clear(array);
                }
```
should be outside   
```
            if (haveBucket)
            {
```
otherwise memory goes unclean and unnoticed.

* In method Bucket.Rent
    `allocateBuffer = buffer == null; `
   is unnecessay inside the spin.
   you can remove the variable allocatedBuffer 
   and change   if (allocateBuffer) => if (buffer == null)
   as buffer is a local variable

**Comment by teo-tsirpanis** [comment, marked as answer]
* If `haveBucket` is false the array will be eligible for collection by the GC. It doesn't matter if it holds references to other objects, the GC will be able to collect them too.
* You are likely right about this one but the difference would be so tiny I doubt it would be even detected by benchmarks.

**Reply by PathogenDavid** [reply]
It's also worth noting that the clear behavior is [explicitly documented](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1.return?view=net-7.0#system-buffers-arraypool-1-return(-0()-system-boolean)):

> If `clearArray` is set to `true`, **and if the pool will store the buffer to enable subsequent reuse**, the [Return(T[], Boolean)](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1.return?view=net-7.0#system-buffers-arraypool-1-return(-0()-system-boolean)) method will clear the `array` [...]
